Ideas for documentation improvements

davidtheclark opened the pull request:
@TrySound I tried to help with documentation by drafting a new README. My goal was just to more clearly and completely document the API and the concepts involved. Please feel free to take whatever suggestions you like and pass on whatever you don't.

TrySound commented:
It's great, thanks a lot. I will check in details after returning from Moscow.

TrySound commented:
For now I think need to specify all props of types. All types have values. Also should be specified all values of div.

TrySound commented:
Let's make slash example with bordr radius or font decl

TrySound commented:
URL has non standard parser. It's regular if first node is string, but else take everything except before and after spaces. You can deep dive in code for more information.
Sorry, I'm from mobile))

davidtheclark commented:
> For now I think need to specify all props of types. All types have values. Also should be specified all values of div.

For some reason I skipped over `value` on a few nodes. Oops. I added those in, along with the `div` values. Was there anything else you had in mind?

> Let's make slash example with bordr radius or font decl

Done.

> URL has non standard parser. It's regular if first node is string, but else take everything except before and after spaces. You can deep dive in code for more information.

I'm not sure I really understand what you're saying here. After looking through the tests I added this sentence to the `word` node: "Unquoted `url()` arguments are also considered word nodes,
e.g. `foo/bar.gif` in `url(foo/bar.gif)`." Does that capture your intention, or was there more to it?

TrySound commented:
@davidtheclark Cool! I meant that everything between parentheses in url function (including spaces, divs and words).is one word. Just to let to use url function content without stringifying.

davidtheclark commented:
@TrySound so if the first argument is a string, you get parsing as in this test: https://github.com/TrySound/postcss-value-parser/blob/master/test/parse.js#L250

But it if is a word you just get a long word, even with spaces ...

Is that right?

I don't understand the justification for that so I'm not sure how to explain it. Could you explain?

TrySound commented:
@davidtheclark Except before and after. They are like in all functions. Will add more tests for this.

TrySound commented:
Here, slash or space, doesn't matter. We just indexOf `)` and then trim the space before it.
https://github.com/TrySound/postcss-value-parser/blob/master/test/parse.js#L300-L306

Here's irregular expection, but only with errored stuff.
https://github.com/TrySound/postcss-value-parser/blob/master/test/parse.js#L241-L248

TrySound commented:
@davidtheclark Here's code part
https://github.com/TrySound/postcss-value-parser/blob/master/lib/parse.js#L97-L132

1. If quotes then parse as usual function else find unescaped `)`.
1. trim whitespace before it.
1. Add node.
1. if function is unclosed add that space as regular space node

ben-eb commented on README.md:
````diff
@@ -24,66 +33,128 @@ var parser = require('postcss-value-parser');
+  }
+]
+```
 
````
Here it would be nice to show some examples of the parsing for `url` functions, as the behaviour is slightly different. Perhaps it might be nice to include this sentence, and an example with a quoted string and one without.

> Additionally, `url` functions are parsed differently depending on whether the first non-space character is a quote. `url("foo/bar/baz.jpg")` is parsed to:
>
> But `url(foo/bar/baz.jpg)` is parsed to:

davidtheclark commented:
I made an attempt at explaining this `url()` difference. To be honest, though, I still don't get it. It seems to do exactly what I expect it to do, so I don't understand what difference needs to be explained. If @TrySound or @ben-eb understand better, feel free to pitch in.

TrySound commented:
@davidtheclark Just for clarity :) Thanks a lot. Now it seems good. Let's squash?

davidtheclark commented:
Squashed.

TrySound commented:
@davidtheclark Seems good.

davidtheclark commented:
Great. Should I merge?

davidtheclark commented:
Glad I could help. Thanks for working on this module.

TrySound commented:
I appreciate your help!:)

TrySound commented:
@davidtheclark We forget to describe `unclosed` functions/strings property :)
